components/vision.py

```diff
@@ -0,0 +1,184 @@
+import math
+from typing import Optional
+
+import wpilib
+import wpiutil.log
+from magicbot import tunable
+from photonlibpy.photonCamera import PhotonCamera  # type: ignore
+from photonlibpy.photonTrackedTarget import PhotonTrackedTarget  # type: ignore
+from wpimath.geometry import Pose2d, Rotation3d, Transform3d, Translation3d, Pose3d
+
+from components.chassis import ChassisComponent
+from utilities.game import apriltag_layout
+
+
+class VisualLocalizer:
+    """
+    This localizes the robot from AprilTags on the field,
+    using information from a single PhotonVision camera.
+    """
+
+    BEST_POSE_BIAS = (
+        1.2  # Give bias to the best pose by multiplying this const to the alt dist
+    )
+
+    add_to_estimator = tunable(False)
```

Comment:
We should probably make this default to true, there is no reason we can't localise before the robot is enabled